examples/nlp/intent_detection_slot_tagging/joint_intent_slot_infer.py

```diff
@@ -103,29 +104,181 @@ def get_preds(logits):
     return np.argmax(logits, 1)
 
 
-intent_logits, slot_logits, loss_mask, subtokens_mask, intents, slot_labels = [
+def log_misclassified_queries(intent_labels, intent_preds, queries, intent_dict, limit=50):
+    logging.info(f'*** Misclassified intent queries (limit {limit}) ***')
+    cnt = 0
+    for i in range(len(intent_preds)):
+        if intent_labels[i] != intent_preds[i]:
+            query = queries[i].split('\t')[0]
+            logging.info(
+                f'{query} - intent: {intent_dict[intent_labels[i]]} - predicted: {intent_dict[intent_preds[i]]}'
+            )
+            cnt = cnt + 1
+            if cnt >= limit:
+                break
+
+
+def log_misclassified_slots(
+    intent_labels, intent_preds, slot_labels, slot_preds, subtokens_mask, queries, intent_dict, slot_dict, limit=50
+):
+    logging.info('')
+    logging.info(f'*** Misclassified slots queries (limit {limit}) ***')
+    # print slot dictionary
+    str = ''
+    for i, slot in enumerate(slot_dict):
+        str += f'{i} - {slot}, '
+        if i % 5 == 4 or i == len(slot_dict) - 1:
+            logging.info(str)
+            str = ''
+
+    logging.info('----------------')
+    cnt = 0
+    for i in range(len(intent_preds)):
+        cur_slot_pred = slot_preds[i][subtokens_mask[i]]
+        cur_slot_label = slot_labels[i][subtokens_mask[i]]
+        if not np.all(cur_slot_pred == cur_slot_label):
+            query = queries[i].split('\t')[0]
+            logging.info(
+                f'{query} - intent: {intent_dict[intent_labels[i]]} - predicted: {intent_dict[intent_preds[i]]}'
+            )
+            logging.info(cur_slot_pred)
+            logging.info(cur_slot_label)
+            cnt = cnt + 1
+            if cnt >= limit:
+                break
+
+
+# check non compliance of B- and I- slots,
+def check_problematic_slots(slot_preds_list, slot_dict):
+    slot_dict = open(slot_dict, 'r').readlines()
+    cnt = 0
+
+    # for sentence in slot_preds:
+    # slots = sentence.split(" ")
+    sentence = slot_preds_list
+    for i in range(len(sentence)):
+        slot_name = slot_dict[int(sentence[i])]
+        if slot_name.startswith("I-"):
+            prev_slot_name = slot_dict[int(sentence[i - 1])]
+            if slot_name[2:] != prev_slot_name[2:]:
+                print("Problem: " + slot_name + " - " + prev_slot_name)
+                cnt += 1
+    print("Total problematic slots: " + str(cnt))
+
+
+def errors_per_class(cm, dict):
+    size = cm.shape[0]
```

Comment:
how general are these functions? can we move some of them to utils folder?

Reply:
Ok, I am moving 2 of the functions that seems to be more general ones to callback_utils file